pkg/registries/registries_test.go

```diff
@@ -0,0 +1,328 @@
+					mirrorSourcePolicy: config.AllowContactingSource,
+				},
+				{
+					source:             "quay.io/openshift-release-dev/ocp-release",
+					mirrorSourcePolicy: config.AllowContactingSource,
+				},
+				{
+					source:             "quay.io/openshift-release-dev/ocp-release",
+					mirrorSourcePolicy: config.AllowContactingSource,
+				},
+			},
+			expectedOutput: []mirrorSet{
+				{
+					source:             "quay.io/openshift-release-dev/ocp-release",
+					mirrorSourcePolicy: config.AllowContactingSource,
+				},
+				{
+					source:             "registry.access.redhat.com/ubi9/ubi-minimal",
+					mirrorSourcePolicy: config.NeverContactSource,
+				},
+			},
+		},
+		{
+			name: "same source and duplicated mirrors, ensuring resolveTags is handled correctly",
+			input: []mirrorSet{
+				{
+					source: "registry.access.redhat.com/ubi9/ubi-minimal",
+					mirrors: []mirror{
+						{"mirror.example.net/image", false},
+						{"mirror.example.com/redhat", false},
+					},
+				},
+				{
+					source: "registry.access.redhat.com/ubi9/ubi-minimal",
+					mirrors: []mirror{
+						{"mirror.example.net/image", false},
+						{"mirror.example.com/redhat", true},
+					},
+				},
+			},
+			expectedOutput: []mirrorSet{
+				{
+					source: "registry.access.redhat.com/ubi9/ubi-minimal",
+					mirrors: []mirror{
+						{"mirror.example.com/redhat", true},
+						{"mirror.example.net/image", false},
+					},
+				},
+			},
+		},
+		{
+			name: "different sources",
+			input: []mirrorSet{
+				{
+					source: "registry.access.redhat.com/ubi9/ubi-minimal",
+					mirrors: []mirror{
+						{"mirror.example.com/redhat", false},
+						{"mirror.example.net/image", false},
+					},
+				},
+				{
+					source: "quay.io/openshift-release-dev/ocp-release",
+					mirrors: []mirror{
+						{"mirror.registry.com:443/ocp/release", false},
+					},
+				},
+			},
+			expectedOutput: []mirrorSet{
+				{
+					source: "quay.io/openshift-release-dev/ocp-release",
+					mirrors: []mirror{
+						{"mirror.registry.com:443/ocp/release", false},
+					},
+				},
+				{
+					source: "registry.access.redhat.com/ubi9/ubi-minimal",
+					mirrors: []mirror{
+						{"mirror.example.com/redhat", false},
+						{"mirror.example.net/image", false},
+					},
+				},
+			},
+		},
+	}
+	for _, test := range testCases {
+		t.Run(test.name, func(t *testing.T) {
+			out := mergeMirrorSets(test.input)
+			assert.True(t, compareMirrorSets(out, test.expectedOutput))
+		})
+	}
+}
+
+// Helper function to compare two slices of mirrorSets for equality.
+func compareMirrorSets(a, b []mirrorSet) bool {
+	if len(a) != len(b) {
+		return false
+	}
+	for i := range a {
+		if a[i].source != b[i].source {
+			return false
+		}
+		if len(a[i].mirrors) != len(b[i].mirrors) {
+			return false
+		}
+		for j := range a[i].mirrors {
+			if a[i].mirrors[j].host != b[i].mirrors[j].host || a[i].mirrors[j].resolveTags != b[i].mirrors[j].resolveTags {
+				return false
+			}
+		}
+	}
+	return true
```

Comment:
Why not assert.Equal? (Uses reflect.DeepEqual under the hood)

Reply:
makes sense, I had thought for some reason assert.Equal wouldn't do a strict checking of slice indices